smart_card_connector_app/src/reader-tracker.js

```diff
@@ -0,0 +1,189 @@
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview This file contains the class used to track readers and their
+ * status based on messages received from the hook setup on PC/SC code inside
+ * NaCl (check readerfactory_nacl.cc for more information).
+ */
+
+goog.provide('GoogleSmartCard.ReaderTracker');
+
+goog.require('GoogleSmartCard.DebugDump');
+goog.require('GoogleSmartCard.Logging');
+goog.require('GoogleSmartCard.TypedMessage');
+goog.require('goog.asserts');
+goog.require('goog.log.Logger');
+goog.require('goog.messaging.AbstractChannel');
+goog.require('goog.object');
+goog.require('goog.structs.Map');
+
+goog.scope(function() {
+
+/** @const */
+var GSC = GoogleSmartCard;
+
+/**
+ * Structure used to store information about the reader.
+ * @param {string} name
+ * @param {string} status
+ * @param {string=} opt_error
+ * @constructor
+ * @struct
+ */
+var ReaderInfo = function(name, status, opt_error) {
+  this.name = name;
+  this.status = status;
+  this.error = opt_error;
+};
+
+/**
+ * This class tracks readers, provides methods to retrieve the list of readers
+ * and get updates on change, and listens to messages received from the hook
+ * living in the NaCl space.
+ * @param {!goog.messaging.AbstractChannel} messageChannel
+ * @param {!goog.log.Logger} parentLogger
+ * @constructor
+ */
+GSC.ReaderTracker = function(messageChannel, parentLogger) {
+  /** @private */
+  this.logger_ = parentLogger;
+
+  messageChannel.registerService(
+      'reader_init_add', this.readerInitAddListener_.bind(this), true);
+  messageChannel.registerService(
+      'reader_finish_add', this.readerFinishAddListener_.bind(this), true);
+  messageChannel.registerService(
+      'reader_remove', this.readerRemoveListener_.bind(this), true);
+
+  /**
+   * @type {!goog.structs.Map.<string, !ReaderInfo>}
+   * @private
+   */
+  this.readers_ = new goog.structs.Map;
+
+  /**
+   * @type {!Array.<!function(!Array.<!ReaderInfo>)>}
+   * @private
+   */
+  this.onUpdateListeners_ = [];
+
+  this.logger_.fine('Initialized');
+};
+
+/** @const */
+var ReaderTracker = GSC.ReaderTracker;
+
+/**
+ * @param {!Object|string} message
+ * @private
+ */
+ReaderTracker.prototype.readerInitAddListener_ = function(message) {
+  var name = message['readerName'];
+  var port = message['port'];
+  var device = message['device'];
+
+  this.logger_.info(
+      'readerInitAddListener_ called for ' + name + ' (port ' + port +
+      ', device ' + device + ')');
+
+  this.readers_.set(port, new ReaderInfo(name, 'yellow'));
+  this.fireOnUpdateListeners_();
+};
+
+/**
+ * @param {!Object|string} message
+ * @private
+ */
+ReaderTracker.prototype.readerFinishAddListener_ = function(message) {
+  var name = message['readerName'];
+  var port = message['port'];
+  var device = message['device'];
+  var returnCode = message['returnCode'];
+  var returnCodeHex = '0x' + (returnCode >>> 0).toString(16);
```

Comment:
GSC.DebugDump.dump would do the same; this would remove the tricky conversions here and would also apply the better formatting (e.g. the usual capital hexadecimal letters, proper length, treating negative numbers in the more expected way, etc.).